### PR TITLE
[WIP] Windows Subsystem for Linux

### PR DESCRIPTION
**Pull request by ghuntley**
# backstory

See https://github.com/NixOS/nixpkgs/issues/30391

# technical information

WSL2 is different to WSL1, see https://docs.microsoft.com/en-us/windows/wsl/wsl2-install for specifics but overall does not change how operating system distributions should package, register and ship their distribution into the app store. 

See below for reference implementations for WSL1 which also work for WSL2:
- https://github.com/microsoft/WSL-DistroLauncher 
- https://salsa.debian.org/rhaist-guest/WSL/blob/master/create-targz.sh
- https://wslstorestorage.blob.core.windows.net/wslblob/Ubuntu.1604.2017.711.0_v1.appx
- https://github.com/Microsoft/WSL/issues/2618
- https://github.com/DDoSolitary/LxRunOffline/releases
- https://github.com/Microsoft/WSL/blob/master/CONTRIBUTING.md#8-detailed-logs

# contacts at microsoft

There's certain discussions I'll be able to have with Microsoft that others won't (NDA info stuffs) but for everything else these two folks are our contacts:

- https://twitter.com/benhillis (software engineer)
- https://twitter.com/craigaloewen (program management)

# next steps
- [x] Registered an account for the NixOS Foundation in the MSFT app store and reserve "NixOS" as an app name. 
- [ ] Identify people to help out (comment on this PR and I'll grant you push access)
- [ ] WSL does not support systemd without https://github.com/arkane-systems/genie so we need to check it out.
- [ ] Create expression that generates an installation of nixos (without grub, containermode?) that produces a `install.tar.gz`
- [ ] Import `https://github.com/microsoft/WSL-DistroLauncher`, customize it and consume `install.tar.gz`.
- [ ] Iterate until it works on a Windows 10 machine
- [ ] Send the `appx` to Microsoft (speak with Geoff) for pre-check verification.
- [ ] How do we make this repeatable ala @GrahamcOfBorg? `WSL-DistroLauncher`  requires windows.
- [ ] Obtain appstore credentials from @edolstra 
- [ ] Automatic (or manual via shell script) uploading of the generated `appx` to the Windows Store

### CODE REVIEW

**Comment by Renha** [comment]
What kind of people are required for helping out?

**Comment by ghuntley** [comment]
@Renha it would be rad to have you help out. Are you able to take on:

> Create expression that generates an installation of nixos (without systemd, without grub, containermode?) that produces a install.tar.gz and/or investigate https://github.com/arkane-systems/genie

**Comment by Rizary** [comment]
I can't wait to have NixOS on my WSL2. I hope I can contribute more to this.

**Comment by ghuntley** [comment]
@Rizary @Renha it would be rad to have you help out. Are you able to take on:

> Create expression that generates an installation of nixos (without systemd, without grub, containermode?) that produces a install.tar.gz and/or investigate https://github.com/arkane-systems/genie

File: nixos/modules/virtualisation/wsl-config.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      dos2unix
+      git-lfs
+      gitAndTools.gitFull
+      htop
+      iotop
+      sudo
+      tmux
+      vim_configurable
+      wget
+      zip
```
Comment (Mic92) [inline nixos/modules/virtualisation/wsl-config.nix]: This list seems rather opinionated. Should the default packages installed in NixOS not already enough?

Comment (ghuntley) [reply nixos/modules/virtualisation/wsl-config.nix]: Indeed they are, that's what I need to be personally productive. When this merges we should ship the default packages.

**Comment by aanderse** [comment]
@ghuntley how is this going? Did @vaibhavsagar and you connect on this after discussing during `officehours`?

**Comment by Rizary** [comment]
> @Rizary @Renha it would be rad to have you help out. Are you able to take on:
> 
> > Create expression that generates an installation of nixos (without systemd, without grub, containermode?) that produces a install.tar.gz and/or investigate https://github.com/arkane-systems/genie

Hi, I will try to look into this after GSoC ended. One thing that I understand is that NixOS is integrated quite tightly with systemd on many levels. I think I am gonna look https://github.com/cloudwatt/nix-container-images as an example.

**Comment by Mic92** [comment]
Why would you want to get rid of systemd? Is WSL2 not now a proper linux kernel that should support it?

**Comment by Mic92** [comment]
Should we close this one and hope somebody else picks it up?

**Comment by Hoverbear** [comment]
I was trying  https://github.com/Trundle/NixOS-WSL and it worked ok, maybe it could be cleaned up?

**Comment by samueldr** [comment]
What? no?!

**Comment by Mic92** [comment]
@samueldr I don't think you merged this PR. This sometime happen when people force push.

**Comment by samueldr** [comment]
(Yeah, I know, I've seen that happen to other people before, but it still is annoying!)

**Comment by ghuntley** [comment]
> (Yeah, I know, I've seen that happen to other people before, but it still is annoying!)

(・o・) Apologies for confusion. I've deleted the WSL branch and am starting again from master as quite a bit of time has elapsed.

**Pull request merged**